File: app/src/main/java/com/fyp/agrifarm/app/youtube/VideoRepository.java
package com.fyp.agrifarm.app.youtube;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.fyp.agrifarm.app.youtube.db.VideoDao;
import com.fyp.agrifarm.db.ViewModelDatabase;
import com.fyp.agrifarm.app.youtube.db.ShortVideo;

import java.util.List;

public class VideoRepository {
    private VideoDao videoDao;
    private LiveData<List<ShortVideo>> videosList;

    public VideoRepository(Application application) {
        videoDao = ViewModelDatabase.getInstance(application).videoDao();
        videosList = videoDao.getAllVideos();
    }

    public void insert(ShortVideo... video) {
        new InsertVideosAsyncTask(videoDao).execute(video);
    }

    public LiveData<List<ShortVideo>> getAllVideos() {
        return videosList;
    }

    private static class InsertVideosAsyncTask extends AsyncTask<ShortVideo, Void, Void> {
        private VideoDao videoDao;

        private InsertVideosAsyncTask(VideoDao videoDao) {
            this.videoDao = videoDao;
        }

        @Override
        protected Void doInBackground(ShortVideo... videos) {

            for(ShortVideo video: videos){
                videoDao.insert(video);
            }
            return null;
        }
    }

}

File: app/src/main/java/com/fyp/agrifarm/app/youtube/VideoRepository.kt
package com.fyp.agrifarm.app.youtube

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.fyp.agrifarm.app.MainActivity
import com.fyp.agrifarm.app.log
import com.fyp.agrifarm.app.youtube.db.ExtendedVideo
import com.fyp.agrifarm.app.youtube.db.ShortVideo
import com.fyp.agrifarm.app.youtube.db.VideoDao
import com.fyp.agrifarm.db.ViewModelDatabase
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import kotlinx.coroutines.*
import java.io.IOException
import kotlin.coroutines.CoroutineContext

class VideoRepository(application: Application) : CoroutineScope {

    private val parentJob = Job()

    override val coroutineContext: CoroutineContext
        get() = parentJob + Dispatchers.IO
    private val scope = CoroutineScope(coroutineContext)

    private val videoDao: VideoDao = ViewModelDatabase.getInstance(application).videoDao()
    val videoList = MutableLiveData<List<ExtendedVideo>>()


    @ExperimentalCoroutinesApi
    fun fetchVideos() {
        // Just to remove the warning (Added Dispatchers.IO again)
        scope.launch(Dispatchers.IO) {
            try {
                val list = YoutubeDataRequest.instance.getDataFromApi()
                videoList.postValue(list)
                videoDao.insert(*list.toTypedArray())
            } catch (e: IOException) {
                loadDbCachedVideos()
                e.printStackTrace()
                cancel("Couldn't complete YouTube request", e)
            }
        }
    }

    private fun loadDbCachedVideos() {
        videoList.postValue(videoDao.allVideos)
    }
}